src/extension/infer.rs

```diff
@@ -0,0 +1,855 @@
+            // arose because one node is a dataflow parent and the other
+            // is one of it's I/O nodes. In that case, the directions could be
+            // the same, and we should try to detect it
+            if dir1 != dir2 && node1 != node2 {
+                let [(src, src_rs), (tgt, tgt_rs)] = if *dir2 == Direction::Incoming {
+                    [(node1, rs1.clone()), (node2, rs2.clone())]
+                } else {
+                    [(node2, rs2.clone()), (node1, rs1.clone())]
+                };
+
+                if src_rs.is_subset(&tgt_rs) {
+                    Some(InferExtensionError::EdgeMismatch(
+                        ExtensionError::TgtExceedsSrcExtensions {
+                            from: *src,
+                            from_offset: Port::new(Direction::Outgoing, 0),
+                            from_extensions: src_rs,
+                            to: *tgt,
+                            to_offset: Port::new(Direction::Incoming, 0),
+                            to_extensions: tgt_rs,
+                        },
+                    ))
+                } else {
+                    Some(InferExtensionError::EdgeMismatch(
+                        ExtensionError::SrcExceedsTgtExtensions {
+                            from: *src,
+                            from_offset: Port::new(Direction::Outgoing, 0),
+                            from_extensions: src_rs,
+                            to: *tgt,
+                            to_offset: Port::new(Direction::Incoming, 0),
+                            to_extensions: tgt_rs,
+                        },
+                    ))
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+        err.unwrap_or(InferExtensionError::MismatchedConcrete {
+            expected: rs1,
+            actual: rs2,
+        })
+    }
+
+    /// Take a group of equal metas and merge them into a new, single meta.
+    ///
+    /// Returns the set of new metas created and the set of metas that were
+    /// merged.
+    fn merge_equal_metas(&mut self) -> Result<(HashSet<Meta>, HashSet<Meta>), InferExtensionError> {
+        let mut merged: HashSet<Meta> = HashSet::new();
+        let mut new_metas: HashSet<Meta> = HashSet::new();
+        for cc in self.eq_graph.ccs().into_iter() {
+            // Within a connected component everything is equal
+            let combined_meta = self.fresh_meta();
+            for m in cc.iter() {
+                if self.shunted.contains_key(m) {
+                    continue;
+                }
+
+                if let Some(cs) = self.constraints.get(m).cloned() {
+                    for c in cs
+                        .into_iter()
+                        .filter(|c| !matches!(c, Constraint::Equal(_)))
+                    {
+                        self.add_constraint(combined_meta, c.clone());
+                    }
```

Comment:
```suggestion
                if let Some(cs) = self.constraints.get(m) {
                    for c in cs
                        .iter()
                        .filter(|c| !matches!(c, Constraint::Equal(_)))
                        .cloned()
                        .collect::<Vec<_>>()
                        .into_iter()
                    {
                        self.add_constraint(combined_meta, c);
                    }
```

Reply:
You need an allocation to avoid borrow checker getting mad but you don't have to
clone the whole thing